server/app/filters/SessionTimeoutFilter.java

```diff
@@ -0,0 +1,105 @@
+package filters;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import auth.CiviFormProfile;
+import auth.ProfileUtils;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Base64;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+import javax.inject.Inject;
+import javax.inject.Provider;
+import org.apache.pekko.stream.Materializer;
+import play.libs.Json;
+import play.mvc.Filter;
+import play.mvc.Http;
+import play.mvc.Result;
+import services.session.SessionTimeoutService;
+import services.settings.SettingsManifest;
+
+/**
+ * A filter to manage session timeouts in CiviForm.
+ *
+ * <p>This filter is responsible for setting a cookie with the current session's timeout data. The
+ * cookie is used by the frontend to determine when to show a warning to the user that their session
+ * is about to expire.
+ */
+public class SessionTimeoutFilter extends Filter {
+  public static final int DEFAULT_INACTIVITY_TIMEOUT_MINUTES = 30;
+  public static final int DEFAULT_MAX_SESSION_DURATION_MINUTES = 600;
+
+  private static final String TIMEOUT_COOKIE_NAME = "session_timeout_data";
+  private static final Duration COOKIE_MAX_AGE = Duration.ofDays(2);
+
+  private final ProfileUtils profileUtils;
+  private final Provider<SettingsManifest> settingsManifest;
+  private final Provider<SessionTimeoutService> sessionTimeoutService;
+
+  @Inject
+  public SessionTimeoutFilter(
+      Materializer mat,
+      ProfileUtils profileUtils,
+      Provider<SettingsManifest> settingsManifest,
+      Provider<SessionTimeoutService> sessionTimeoutService) {
+    super(mat);
+    this.profileUtils = checkNotNull(profileUtils);
+    this.settingsManifest = checkNotNull(settingsManifest);
+    this.sessionTimeoutService = sessionTimeoutService;
+  }
+
+  @Override
+  public CompletionStage<Result> apply(
+      Function<Http.RequestHeader, CompletionStage<Result>> nextFilter,
+      Http.RequestHeader requestHeader) {
+
+    Optional<CiviFormProfile> optionalProfile =
+        profileUtils.optionalCurrentUserProfile(requestHeader);
+    if (!settingsManifest.get().getSessionTimeoutEnabled() || optionalProfile.isEmpty()) {
+      return nextFilter.apply(requestHeader).thenApply(this::clearTimeoutCookie);
+    }
+    return nextFilter
+        .apply(requestHeader)
+        .thenCompose(
+            result ->
+                createSessionTimestampCookie(optionalProfile.get()).thenApply(result::withCookies));
+  }
+
+  private CompletableFuture<Http.Cookie> createSessionTimestampCookie(CiviFormProfile profile) {
```

Comment:
```suggestion
  /**
   * This cookie is used by client-side Javascript. It is not encrypted because the client needs to 
   * read it. We should avoid using values in the cookie that are sent on client requests. Extra
   * validation should be performed if we need to use the request cookie on the server.
   */
  private CompletableFuture<Http.Cookie> createSessionTimestampCookie(CiviFormProfile profile) {
```

(blocking) Adding a note on the usage of the cookie. Please reword if needed.